docs/proposals/cross-tenant-query-federation.md

```diff
@@ -0,0 +1,82 @@
+---
+title: "Cross-Tenant Query Federation"
+linkTitle: "Cross-Tenant Query Federation"
+weight: 1
+slug: "cross-tenant-query-federation"
+---
+
+- Author: [Christian Simon](https://github.com/simonswine)
+- Date: October 2020
+- Status: Proposed
+
+## Overview
+
+This document aims to describe how to implement the ability to allow queries to cover data from more than a single Cortex tenant.
+
+## Reasoning
+
+Adopting a tenancy model within an organization with each tenant representing a department comes with the disadvantage that it will prevent queries from spanning multiple departments. This proposal tries to overcome those limitations.
+
+## Alternatives considered
+
+In theory PromQL API clients could be aggregating/correlating query results from multiple tenants. For example Grafana could be used with multiple data sources and a cross tenant query could be achieved through using [Transformations][grafana_transformation].
+
+[grafana_transformation]: https://grafana.com/docs/grafana/latest/panels/transformations/
+
+## Challenges
+
+### Aggregate data without overlaps
+
+#### Challenge
+
+The series in different tenants might have exactly the same labels and hence potentially collide which each other.
+
+#### Proposal
+
+In order to identify the series uniquely, queries using multiple tenants should inject a label with the name `__tenant__` and its value containing the tenant ID into the results. An potentially existing label with the same name should be overwritten.
```

Comment:
👍 I like the approach of using a reserved label.

Reply:
Could the injected label be configurable? Depending on the organization, I imagine tenants could be used similarly to external_labels where all data for certain environments or products are split into different tenants. Then the label could be much more meaningful to a user than `__tenant__`.

The behavior where potentially existing labels being overwritten is interesting, the exact opposite approach is taken by Prometheus default remote behavior - if the label already exists in the series it is not overwritten by `external_labels`. There is also the behavior of scraping which can add `exported_<original-label-name>`. Personally I would prefer one of those approaches than overwriting the label but am not set strongly.

Reply:
@csmarchbanks yes a very good point. The injected label should be configurable. 

I think the tenant_id label value needs to take precedence over what is stored on the series, as the tenant isolation is IMHO much more important, than preserving the user's label data. (Thanos also seems to do this overwrite without keeping the original value: https://github.com/thanos-io/thanos/blob/4b89eea19e6c477e5d272aab9db6b46afee8d2bb/pkg/store/labelpb/label.go#L242)

I do agree that it might be helpful to preserve it in a similar way that the prefix `exporter_` works and it should be something relatively  straightforward to add to the proposal. What do you think should we do with this:

- Is `exporter_` the right prefix, or should we do something like `overwritten_by_tenant_` or maybe `original_`?

- Do we need to enforce that recursively let's say our prefix is `original_` what are we gonna do in the case when the input contains `__tenant__` and `original___tenant__`?

Reply:
I think `original_` is nice, and would not overlap with `exported_` from scrapes. I guess `original_` could become `original_original_...` or something in the case of multiple matches (though that seems very unlikely).

Considering that this is the display layer we are talking about, not actually accessing the data, I am not sure I am convinced that the tenant info is more important than being able to easily transport queries from say a local Prometheus to Cortex. I could also understand an argument where one would wish that how tenants are structured in Cortex is transparent from a user's perspective.

Reply:
Also, would it be possible to query on `__tenant__`? It could be unexpected to get a label back in a result set that would not be queryable.

Reply:
> Also, would it be possible to query on `__tenant__`? It could be unexpected to get a label back in a result set that would not be queryable.

Yes it will be possible to query, the labels matchers are applied to figure out which tenants to use

> Considering that this is the display layer we are talking about, not actually accessing the data, I am not sure I am convinced that the tenant info is more important than being able to easily transport queries from say a local Prometheus to Cortex. I could also understand an argument where one would wish that how tenants are structured in Cortex is transparent from a user's perspective.

I still think it is very important to display the accurate tenant_id in `__tenant__` above anything else. 

Just a quick example: you have a manager A and two teams B, C each with their own tenant. The boss evaluates the performance of the teams through a multi-tenant query. If the raw data takes precedence team C could manipulate some data to make team B look bad just by using the `__tenant__` label.

Reply:
Your use case is different from mine :) I sure hope I never have to deal with that situation.

Cool, I think `original_` is a good approach then and make sure there is a tenant label for everything. If someone wants they can aggregate it away anyway.